Functions.scala

```diff
@@ -0,0 +1,163 @@
+class Functions extends App {
+  //Helpers
+  //time complexity O(n), n - list.length
+  //space complexity O(1)
+  private def rev[A](list: List[A]): List[A] = {
+    @scala.annotation.tailrec
+    def revIter[A](listToRev: List[A], list: List[A]): List[A] = {
+      listToRev match {
+        case Nil => list
+        case _ => revIter(listToRev.tail, listToRev.head :: list)
+      }
+    }
+
+    revIter(list, List())
+  }
+
+  //exercise 1
+  //time complexity O(n),  n = listToMerge.length
+  //space complexity O(1) - tail recursion, optimizes stack frames, gets rid of extra calls
+  def split(listToMerge: List[Int]): (List[Int], List[Int]) = {
+    @scala.annotation.tailrec
+    def splitIter(source: List[Int], first: List[Int], second: List[Int]): (List[Int], List[Int]) = {
+      if (source == Nil) (rev(first), rev(second))
+      else if (source.head < 0)
+        if (source.head % 2 != 0)
+          splitIter(source.tail, source.head :: first, source.head :: second)
+        else splitIter(source.tail, source.head :: first, second)
+      else splitIter(source.tail, first, second)
+    }
+
+    splitIter(listToMerge, List(), List())
+  }
+
+  //exercise 2
+  //time complexity O(n),  n = list.length
+  //space complexity O(1)
+  def length[A](list: List[A]): Int = {
+    @scala.annotation.tailrec
+    def lengthIter[A](list: List[A], length: Int): Int = {
+      list match {
+        case Nil => length
+        case _ => lengthIter(list.tail, length + 1)
+      }
+    }
+
+    lengthIter(list, 0)
+  }
+
+  //exercise 3
+  //time complexity O(n+m), n = first.length, m = second.length
```

Comment:
nein,nein,nein! ;) time complexity is the length of a shorter list. When a shorter list ends you just append the remaining elems.